core/src/main/java/jenkins/console/ConsoleUrlProvider.java

```diff
@@ -0,0 +1,121 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.console;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Functions;
+import hudson.model.Describable;
+import hudson.model.Descriptor;
+import hudson.model.Run;
+import hudson.model.User;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import jenkins.model.Jenkins;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.Stapler;
+
+/**
+ * Extension point that allows implementations to redirect build console links to a specified URL.
+ * <p>In order to produce links to console URLs in Jelly templates, use {@link Functions#getConsoleUrl}.
+ * @see Functions#getConsoleUrl
+ * @since TODO
+ */
+public interface ConsoleUrlProvider extends Describable<ConsoleUrlProvider> {
+    @Restricted(NoExternalUse.class)
+    Logger LOGGER = Logger.getLogger(ConsoleUrlProvider.class.getName());
+
+    /**
+     * Get a URL relative to the context path of Jenkins which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the context of Jenkins, or {@code null}
+     * if this implementation does not want to server a special console view for this build.
+     */
+    @CheckForNull String getConsoleUrl(Run<?, ?> run);
+
+    @Override
+    default Descriptor<ConsoleUrlProvider> getDescriptor() {
+        return Jenkins.get().getDescriptorOrDie(getClass());
+    }
+
+    /**
+     * Get a URL relative to the web server root which should be used to link to the console for the specified build.
+     * <p>Should only be used in the context of serving an HTTP request.
+     * <p>Use {@link Functions#getConsoleUrl} to obtain this link in a Jelly template.
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the web server root
+     */
+    static @NonNull String getRedirectUrl(Run<?, ?> run) {
+        ConsoleUrlProviderGlobalConfiguration globalConfig = ConsoleUrlProviderGlobalConfiguration.get();
+        List<ConsoleUrlProvider> providers = globalConfig.getProviders();
+        User currentUser = User.current();
+        if (currentUser != null) { // TODO: Should we add a configuration option to allow admins to suppress this behavior?
+            ConsoleUrlProviderUserProperty userProperty = currentUser.getProperty(ConsoleUrlProviderUserProperty.class);
+            if (userProperty != null) {
+                List<ConsoleUrlProvider> userProviders = userProperty.getProviders();
+                if (userProviders != null) {
+                    providers = userProviders;
+                }
```

Comment:
This looks like a user who configures a provider that does return `null` will not fall back to the admin choice, but to the default provider. Is that intentional?

Reply:
I updated the behavior and added a test in [b5420ad](https://github.com/jenkinsci/jenkins/pull/8321/commits/b5420ad07c46402632bdd712a5cac7ffb88bda04).